proposals/0031-hlsl-vector-matrix-operations.md

```diff
@@ -156,12 +144,500 @@ vector<float, 3> ApplyNeuralMaterial(vector<half, 8> inputVector) {
+  DATA_TYPE_FLOAT32 = 9,          // ComponentType::F32
+  DATA_TYPE_SINT8_T4_PACKED = 17, // ComponentType::PackedS8x32
+  DATA_TYPE_UINT8_T4_PACKED = 18, // ComponentType::PackedU8x32
+  DATA_TYPE_UINT8 = 19,           // ComponentType::U8
+  DATA_TYPE_SINT8 = 20,           // ComponentType::I8
+  DATA_TYPE_E4M3 = 21,            // ComponentType::F8_E4M3
+                                  // (1 sign, 4 exp, 3 mantissa bits)
+  DATA_TYPE_E5M2 = 22,            // ComponentType::F8_E5M2
+                                  // (1 sign, 5 exp, 2 mantissa bits)
+};
+```
+
+See the Type Interpretations section of [0029] for more information.
+
+### enum MatrixLayout
+
+This enum defines the different possible layouts of a matrix in memory.
+
+```c++
+enum MatrixLayout {
+  MATRIX_LAYOUT_ROW_MAJOR = 0,
+  MATRIX_LAYOUT_COLUMN_MAJOR = 1,
+  MATRIX_LAYOUT_MUL_OPTIMAL = 2,
+  MATRIX_LAYOUT_OUTER_PRODUCT_OPTIMAL = 3
+};
+```
+
+See the Matrix Layouts section of [0029] for more information.
+
+### struct MatrixRef / RWMatrixRef
+
+`MatrixRef` and `RWMatrixRef` specify a reference to a matrix in memory,
+including information about the dimensions, layout and numerical format of the
+matrix. Some of this information must be known at compile time - these are
+captured in template parameters - while others can be determined at runtime, and
+are therefore members of the struct.
+
+Since the `ByteAddressBuffer` and `RWByteAddressBuffer` are not convertable to
+each other, use use different named types for the matrix references. A base
+class, `MatrixRefImpl` is used for the common code.
+
+Other functions in the API are implemented in terms of `MatrixRefImpl`. This
+allows the compiler to generate more useful diagnostics if wrong types are used.
+
+Example usage:
+
+```c++
+ByteAddressBuffer ROBuffer;
+RWByteAddressBuffer RWBuffer;
+
+void Example() {
+  using namespace dx::linalg;
+
+  MatrixRef<DATA_TYPE_FLOAT16, 4, 4, MATRIX_LAYOUT_MUL_OPTIMAL, true> MatrixA =
+      {ROBuffer, /*offset=*/128, /*stride=*/0};
+
+  MatrixRef<DATA_TYPE_FLOAT16, 4, 4, MATRIX_LAYOUT_ROW_MAJOR, true>
+      MatrixB = {ROBuffer, /*offset=*/128, /*stride=*/16};
+
+  RWMatrixRef<DATA_TYPE_FLOAT16, 128, 256, MATRIX_LAYOUT_OUTER_PRODUCT_OPTIMAL>
+      MatrixC = {RWBuffer, /*offset=*/64, /*stride=*/0};
+}
+```
+
+Template parameters:
+
+- `DT` - the type used to store elements of the matrix in memory
+- `M` - the 'M' dimension of the matrix
+- `K` - the 'K" dimension of the matrix
+- `ML` - the layout of the matrix in memory
+- `Transpose` - whether or not this matrix should be transposed before any
+  operations
+
+Members:
+
+- `Buffer` - the buffer that the matrix is stored in
+  - For `MatrixRef` this is a `ByteAddressBuffer`
+  - For `RWMatrixRef` this is a `RWByteAddresssBuffer`
+- `StartOffset` - the offset, in bytes, from the beginning of the buffer where
+  the matrix is located.
+- `Stride` - the stride, in bytes, between rows/columns of the matrix. This
```

Comment:
Minor nit: I'm sure the audience of this spec would be clear on this. But saying 'rows or columns' is more explicit and correct here vs 'rows/columns'.

Reply:
Done